nipasswd.8
.\" $Id: nipasswd.8,v 1.6 2002/08/24 07:15:35 chip Exp $
.TH NIPASSWD 8L
.SH NAME
nipasswd - Non-interactive password utility for PAM systems.
.SH SYNTAX
.B nipasswd
[
.B \-Dac
]
.SH DESCRIPTION
The
.I nipasswd
utility is a non-interactive program to authenticate users or change
account passwords.  It is intended to provide these capabilities--normally
privileged operations--to other scripts and programs in a moderately
safe and secure fashion.
.P
By default,
.I nipasswd
attempts to change the password on a user account.  At startup, it reads
three lines of input from stdin:
.RS
.P
.nf
username
old_password
new_password
.fi
.RE
.P
It first verifies that the
.I old_password
is correct for the
.I username
account, and then attempts to change it to
.IR new_password .
.P
.P
If the
.B \-a
option is given, then
.I nipasswd
changes to authenticate mode.  In this mode, it reads just two lines
from stdin (\fIusername\fP and \fIpassword\fP) and attempts to authenticate
the user.
.P
The
.B \-D
option enables debug output.
.P
A minimum UID is enforced, as controlled by the MIN_AUTH_UID and
MIN_CHANGE_UID values compiled into the program.  An attempt to authorize
a user or change a password for an account below these limits will
be denied.  In the distributed version of this program, both values are
set to 100.
.P
This utility operates under
the
.IR pam (8)
system, using a service name of "nipasswd".
.SH EXIT STATUS
.IP 0 4
Password has been successfully changed.
.IP 1 4
Failure due to some error.  Password unchanged.
.IP 2 4
Failure due to authorization error.  Either the
.I username
does not exist, the
.I old_password
is incorrect, or the UID of this
account below the minimum permitted.  Password unchanged.
.IP 3 4
Failure due to bad password checks.  This will occur only when the
.B \-c
option is given.  Password unchanged.
.P
In the case of non-zero exit status, a diagnostic message will be
written to the output.
.SH FILES
/etc/pam.d/nipass
.SH BUGS
This utility does not implement any guards against dictionary attacks on
account passwords, other than invoking a delay penalty on authentication
failures.
.P
The PAM API makes my head hurt.
.SH AUTHOR
.nf
This file is part of the "web-chpass" package.
See <https://github.com/chip-rosenthal/web-chpass> for more information.

Chip Rosenthal
<chip@unicom.com>
.fi
.SH SEE ALSO
pam(8), passwd(1)


CHANGES
Fri Mar  4 10:21:55 CST 2016

        Fixed nipasswd so it runs "obscure" checks correctly.

        Dropped ENABLE_BADPW_CHECK configuration setting. Bad password
        check should always be run.

        Dropped FORCE_HTTPS configuration setting and make secure
        connection required.

        Converted INSTALL to INSTALL.md

        Normalized file headers.

Thu Mar  3 10:37:58 CST 2016

	Added FORCE_HTTPS configuration setting to force application
	to run on a secure connection.

	Added CGIEXT configuration setting to add an optional suffix
	(like ".cgi") to the installed CGI script file.

	Renamed getline() function to avoid lib conflict on build.

	Now uses Perl Template Toolkit to render HTML. Template
	file now called chpass.tmpl.

	Rewrote README and INSTALL.

	Moved to "unlicense" (public domain) and added LICENSE file.

	Moved distribution point to github.

	Tested with a modern Linux (Debian 8 Jessie) and in an LDAP
	environment.

Version 1.4 - 21-Dec-2006

	Fix "Unexpected PAM message (PAM_TEXT_INFO): Password
	changed." problem reported on SuSE 10.1 by Sven Sohnel.

Version 1.3 - 12-Nov-2006

	Give useful message when password has expired (PAM_NEW_AUTHTOK_REQD).

	Resolve Fedora 5 portability issues reported by
	Steve Mokris <smokris@softpixel.com>.  (Thanks for the patch.)

version 1.2 - 15-Jun-2003

	No longer ship a pam.d/nipasswd file.  The one that used to be
	here only worked on Red Hat Linux.  Instead, on installation,
	we'll make it by copying the existing PAM module for the "passwd"
	utility.

Version 1.1 - 24-Aug-2002

	Added BADPW_CHECKS configuration setting, to allow the bad
	password checks to be enabled, disabled, or optional.  Added
	MIN_AUTH_UID and MIN_CHANGE_UID configuration settings, so that
	new min UID feature can be used without hacking the source code.

	Thanks to Dan Senie for the useful suggestions.

Version 1.0 - 24-Aug-2002

	Added minimum UID checks, to avoid accessing/changing system
	accounts.

Version 0.1 - 20-Jul-2002

	Initial release for public testing.

# $Id: CHANGES,v 1.5 2006/12/21 16:36:32 chip Exp $
# $Source: /home/chip/src/web-chpass/RCS/CHANGES,v $


nipasswd.c
/*
 * nipasswd.c - Non-interactive password utility.
 *
 * Part of the "web-chpass" package.
 * https://github.com/chip-rosenthal/web-chpass
 *
 * Chip Rosenthal
 * <chip@unicom.com>
 */

#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <stdarg.h>
#include <pwd.h>
#include <errno.h>
#include <assert.h>
#include <security/pam_appl.h>

#define USAGE			"usage: %s [-Da]"
#define NIPASSWD_PAM_SERVICE	"nipasswd"
#define FAIL_DELAY		5 /* secs */
#define BUFLEN			512

#ifndef MIN_AUTH_UID
# define MIN_AUTH_UID	100	/* do not auth users below this uid	*/
#endif
#ifndef MIN_CHANGE_UID
# define MIN_CHANGE_UID	100	/* do not change users below this uid	*/
#endif

#define EX_SUCCESS	0	/* password successfully changed	*/
#define EX_ERROR	1	/* failed due to an error		*/
#define EX_DENIED	2	/* failed due to username/password auth	*/
#define EX_BADPW	3	/* failed due to bad password checks	*/

#define Dprintf	if (!Debug) ; else fprintf

int Debug = 0;			/* enable debugging messages		*/
int Do_auth_only = 0;		/* authenticate but don't change passwd	*/

/*
 * This information is global so it can be accessed by die().
 */
pam_handle_t *pam_h = NULL;
int pam_rc;

/*
 * This information is global so it can be accessed by pam_conv_func().
 */
int pam_conv_resp_count = 0;
char username[BUFLEN];
char old_password[BUFLEN];
char new_password[BUFLEN];

int user_ok(const char *username, struct passwd *save_pw);
int pam_conv_func(int num_msg, const struct pam_message **msg, struct pam_response **resp, void *appdata_ptr);
void fgetline(char *buf, size_t buflen, FILE *fp);
char *xstrdup(const char *s);
void die(int exitstat, const char *fmt, ...);
const char *pam_msg_style_str(int n);


int main(int argc, char *argv[])
{
	struct pam_conv pam_conv = { pam_conv_func, NULL };
	int i;

	while ((i = getopt(argc, argv, "Da")) != EOF) {
		switch (i) {
		case 'D':
			Debug = 1;
			break;
		case 'a':
			Do_auth_only = 1;
			break;
		default:
			die(EX_ERROR, USAGE, argv[0]);
		}
	}
	if (argc-optind != 0) {
		die(EX_ERROR, USAGE, argv[0]);
	}

	fgetline(username, sizeof(username), stdin);
	fgetline(old_password, sizeof(old_password), stdin);
	if (!Do_auth_only) {
		fgetline(new_password, sizeof(new_password), stdin);
	}
	if (getc(stdin) != EOF) {
		die(EX_ERROR, "Excess input.");
	}

	struct passwd pw;
	if (!user_ok(username, &pw)) {
		(void) sleep(FAIL_DELAY);
		die(EX_DENIED, "Access denied.");
	}

	/*
	 * Make sure our real UID matches the account that's being changed.
	 * On Debian with pam_ldap, this is required else there is
	 * a prompt for the LDAP admin passwd.
	 */
	if (setreuid(pw.pw_uid, 0) != 0) {
		die(EX_ERROR, "Cannot set user id: %m");
	}

	pam_rc = pam_start(NIPASSWD_PAM_SERVICE, username, &pam_conv, &pam_h);
	if (pam_rc != PAM_SUCCESS) {
		die(EX_ERROR, "Error initializing PAM subsystem: %s",
			pam_strerror(pam_h, pam_rc));
	}
#ifdef PAM_FAIL_DELAY
	pam_fail_delay(pam_h, FAIL_DELAY*1000);
#endif

	if (Do_auth_only) {

                /*
                 * Attempt to authenticate the user.
                 */
                pam_rc = pam_authenticate(pam_h, 0);
                switch (pam_rc) {
                case PAM_USER_UNKNOWN:
                case PAM_AUTH_ERR:
                        die(EX_DENIED, "Access denied.");
                case PAM_SUCCESS:
                        break;
                default:
                        die(EX_ERROR, "PAM error authenticating user: %s",
                                pam_strerror(pam_h, pam_rc));
                }

        } else {

                /*
                 * Attempt to change the password.
                 */
                pam_rc = pam_chauthtok(pam_h, 0);
                if (pam_rc != PAM_SUCCESS) {
                        die(EX_ERROR, "Error setting new password: %s:",
                                pam_strerror(pam_h, pam_rc));
                }

        }

	Dprintf(stderr, "main: terminating with success exit status\n");
	(void) pam_end(pam_h, PAM_SUCCESS);
	exit(EX_SUCCESS);
}


/*
 * user_ok() - Verify it is alright to handle this user.
 *
 * The main purpose of this procedure is to enforce the min UID checks.
 * But so long as we are at it, we can bounce unknown users without the
 * overhead of stoking up PAM.
 */
int user_ok(const char *username, struct passwd *save_pw)
{
	struct passwd *pw;
	if ((pw = getpwnam(username)) == NULL) {
		return 0;
	}
	if (pw->pw_uid < (Do_auth_only ? MIN_AUTH_UID : MIN_CHANGE_UID)) {
		return 0;
	}
	if (save_pw != NULL) {
		*save_pw = *pw;
	}
	return 1;
}


int pam_conv_func(int num_msg, const struct pam_message **msg, struct pam_response **resp, void *appdata_ptr)
{
	struct pam_response *resp_buf;
	int handled_this, i;

	Dprintf(stderr,
		"pam_conv_func: entered, pam_conv_resp_count=%d num_msg=%d\n",
		pam_conv_resp_count, num_msg);
	resp_buf = calloc(num_msg, sizeof(struct pam_response));
	if (resp_buf == NULL) {
		die(EX_ERROR, "System error:  malloc failed.");
	}

	for (i = 0 ; i < num_msg ; ++i) {

		handled_this = -1;
		resp_buf[i].resp = NULL;
		resp_buf[i].resp_retcode = PAM_SUCCESS;

		/*
		 * pam_conv_resp_count counts the number of prompted
		 * responses we've sent back to PAM.  There should
		 * be three:  old password, new password, and confirm
		 * new password.
		 */
		switch (pam_conv_resp_count) {

		case 0:	/* prompt for old password */
			switch (msg[i]->msg_style) {

			case PAM_PROMPT_ECHO_OFF:
				resp_buf[i].resp = xstrdup(old_password);
				++pam_conv_resp_count;
				handled_this = 1;
				break;

			case PAM_TEXT_INFO:
				/* "Changing password for ..." */
				handled_this = 1;
				break;

			default:
				handled_this = 0;
				break;

			}
			break;

		case 1: /* prompt for new password */
			switch (msg[i]->msg_style) {

			case PAM_PROMPT_ECHO_OFF:
				resp_buf[i].resp = xstrdup(new_password);
				++pam_conv_resp_count;
				handled_this = 1;
				break;

			case PAM_TEXT_INFO:
				/* "Changing password for ..." */
				handled_this = 1;
				break;

#define EXPMSSG "You are required to change your password immediately"

			case PAM_ERROR_MSG:
				if (strncmp(msg[i]->msg, EXPMSSG, strlen(EXPMSSG)) == 0) {
					handled_this = 1;
					break;
				}
				/* else fall thru */

			default:
				handled_this = 0;
				break;

			}
			break;

		case 2: /* confirm new password */
			switch (msg[i]->msg_style) {

			case PAM_PROMPT_ECHO_OFF:
				resp_buf[i].resp = xstrdup(new_password);
				++pam_conv_resp_count;
				handled_this = 1;
				break;

			case PAM_ERROR_MSG:
                                pam_rc = PAM_PERM_DENIED;
                                die(EX_BADPW, "%s", msg[i]->msg);
				break;

			default:
				handled_this = 0;
				break;

			}
			break;

		case 3: /* possible additional message */
			switch (msg[i]->msg_style) {

			case PAM_TEXT_INFO:
				/* "Password changed." */
				handled_this = 1;
				break;

			case PAM_ERROR_MSG:
                                pam_rc = PAM_PERM_DENIED;
                                die(EX_BADPW, "%s", msg[i]->msg);
				break;

			default:
				handled_this = 0;
				break;

			}
			break;

		default: /* don't know what this is */
			handled_this = 0;
			break;

		}

		assert(handled_this >= 0);
		Dprintf(stderr,
			"pam_conv_func: msg_style=\"%s\" msg=\"%s\" resp=\"%s\"\n",
			pam_msg_style_str(msg[i]->msg_style),
			msg[i]->msg, resp_buf[i].resp);

		if (!handled_this) {
			die(EX_ERROR, "System Error - Unexpected PAM message (%s): %s",
				pam_msg_style_str(msg[i]->msg_style), msg[i]->msg);
		}

	}

	*resp = resp_buf;
	return PAM_SUCCESS;
}


void fgetline(char *buf, size_t buflen, FILE *fp)
{
	int n;

	if (fgets(buf, buflen, fp) == NULL) {
		if (feof(fp)) {
			die(EX_ERROR, "premature end of input");
		} else {
			die(EX_ERROR, "error reading input: %m");
		}
	}

	n = strlen(buf);
	if (n > 0 && buf[n-1] != '\n') {
		die(EX_ERROR, "input rejected: buffer overflow");
	}
	buf[n-1] = '\0';
}


char *xstrdup(const char *s)
{
	char *s1;
	if ((s1 = strdup(s)) == NULL) {
		die(EX_ERROR, "System error:  malloc failed.");
	}
	return s1;
}


void die(int exitstat, const char *fmt, ...)
{
	va_list ap;
	int save_errno = errno;
	char mssgbuf[1024], *s;

	va_start(ap, fmt);
	vsnprintf(mssgbuf, sizeof(mssgbuf), fmt, ap);
	if ((s = strstr(mssgbuf, "%m")) != NULL)  {
		*s = '\0';
		s += 2;
		fputs(mssgbuf, stderr);
		fputs(strerror(save_errno), stderr);
		fputs(s, stderr);
	} else {
		fputs(mssgbuf, stderr);
	}
	putc('\n', stderr);

	if (pam_h != NULL) {
		Dprintf(stderr, "die: terminating with PAM status: %s\n",
			pam_strerror(pam_h, pam_rc));
		(void) pam_end(pam_h, pam_rc);
		pam_h = NULL;
	}
	exit(exitstat);
	/*NOTREACHED*/
}


const char *pam_msg_style_str(int n)
{
	static char smbuf[64];
	switch (n) {
	case PAM_PROMPT_ECHO_OFF:
		return "PAM_PROMPT_ECHO_OFF";
	case PAM_PROMPT_ECHO_ON:
		return "PAM_PROMPT_ECHO_ON";
	case PAM_ERROR_MSG:
		return "PAM_ERROR_MSG";
	case PAM_TEXT_INFO:
		return "PAM_TEXT_INFO";
	default:
		sprintf(smbuf, "<code %d>", n);
		return smbuf;
	}
}
